### PR TITLE
Add pupnp/1.12.1

### PR DESCRIPTION
**Pull request by ytimenkov**
Specify library name and version:  **lib/1.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/pupnp/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    _autotools = None
+
+    _pupnp_libs = ["upnp", "ixml"]
+
+    def configure(self):
+        if self.settings.os == "Windows":
+            # Note, pupnp has build instructions for Windows but they
+            # include VC 6 and require pthreads-w32 library.
+            # Someone who needs it and has possibility to build it could step in.
+            raise ConanInvalidConfiguration("Windows builds are not supported.")
```
Comment (madebr) [inline recipes/pupnp/all/conanfile.py:50]: Good comment! :100:

File: recipes/pupnp/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        for lib in self._pupnp_libs:
+            os.unlink(os.path.join(self.package_folder, "lib", "lib%s.la" % lib))
+
+    def package_info(self):
+        self.cpp_info.libs.extend(self._pupnp_libs)
+        if self.settings.os == "Linux":
+            self.cpp_info.system_libs.extend(["pthread"])
+            self.cpp_info.cflags.extend(["-pthread"])
+            self.cpp_info.cxxflags.extend(["-pthread"])
+        self.cpp_info.includedirs = ["include/upnp"]
```
Comment (madebr) [inline recipes/pupnp/all/conanfile.py:109]:
```suggestion
        self.cpp_info.includedirs.append(os.path.join("include", "upnp"))
```

Comment (ytimenkov) [reply recipes/pupnp/all/conanfile.py:109]: Thanks for review.
I'm not sure if it's a good idea to keep both paths (default is just `["include"]` iirc).

I'm don't know why upnp puts everything into a subfolder but requires it to be in the include path 🤦‍♂️

Generally the amount of customizations and deletes makes me wonder if it's a good idea to rely on `make install` or just copy all files by glob...

Comment (ytimenkov) [reply recipes/pupnp/all/conanfile.py]: This is amt Posix-only, but I agree that `path.join` is furure-proof.
However I think having a subdir is not needed anyways so here is a "clever" solution.

If someone thinks that "clever:" is not good (I have doubts myself) I can change back to moving and removing files.

File: recipes/pupnp/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    def _configure_autotools(self):
+        if not self._autotools:
+            args = [
+                "--enable-static=%s" % ("no" if self.options.shared else "yes"),
+                "--enable-shared=%s" % ("yes" if self.options.shared else "no"),
+                "--disable-samples",
+            ]
+
+            def enable_disable(opt):
```
Comment (uilianries) [inline recipes/pupnp/all/conanfile.py:67]: nice trick

**Pull request merged**